cmd/snap-update-ns/bootstrap.c

```diff
@@ -486,14 +483,15 @@ void bootstrap(int argc, char **argv, char **envp)
 	const char *snap_name = NULL;
 	bool should_setns = false;
 	bool process_user_fstab = false;
-	unsigned long uid = 0;
+	unsigned long uid = getuid();
 	process_arguments(argc, argv, &snap_name, &should_setns,
 			  &process_user_fstab, &uid);
-	if (process_user_fstab) {
-		switch_to_privileged_user();
-		// switch_to_privileged_user sets bootstrap_{errno,msg}
-	} else if (snap_name != NULL && should_setns) {
-		setns_into_snap(snap_name);
+	if (should_setns && snap_name != NULL) {
+		setns_into_snap(snap_name, process_user_fstab ? uid : 0);
 		// setns_into_snap sets bootstrap_{errno,msg}
 	}
+	if (process_user_fstab && uid != 0) {
+		switch_to_specific_privileged_user(uid, getgid());
+		// switch_to_privileged_user sets bootstrap_{errno,msg}
+	}
```

Comment:
Maybe it's just me, but the combination of processing the args in a different file, this and the previous conditional and using the ternary operator here makes this all harder to understand than it could be. That said, the code apears to be correct. I don't have a better suggestion to make it easier to understand, which might be an indication of how snap-update-ns' invocations are too complex... (I can't help but think this could be made simpler... it's difficult to verify the behavior).

Reply:
Btw the name switch_to_privileged_user is a bit strange, the user is not privileged, is that we switch to it but keep/reacquire some our caps afaiu

Reply:
I renamed the function now.

Reply:
@jdstrand I have a plan to simplify this but I want to complete the several ongoing changes that affect this area first. The plan will involve a new binary, `snap-boostrap-ns` that will both take over responsibility of parts of `snap-confine` and also simplify command line handling of `snap-update-ns`, which will now only be invoked by `snapd` for actual updates.